gobblin-core/src/main/java/gobblin/writer/BaseDataWriter.java

```diff
@@ -0,0 +1,85 @@
+/* (c) 2015 NerdWallet All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.writer;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.util.ForkOperatorUtils;
+import gobblin.util.WriterUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+
+/**
+ * An implementation of {@link DataWriter} does the work of setting the output/staging dir
+ * and creating the FileSystem instance.
+ *
+ * @author akshay@nerdwallet.com
+ */
+public abstract class BaseDataWriter<D> implements DataWriter<D> {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseDataWriter.class);
+
+  protected final FileSystem fs;
+  protected final Path stagingFile;
+  protected final Path outputFile;
+  protected final State properties;
+
+  public BaseDataWriter(State properties, String fileName, int numBranches, int branchId) throws IOException {
+    this.properties = properties;
+    // initialize file system
+    String uri = properties
+            .getProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, numBranches, branchId),
+                    ConfigurationKeys.LOCAL_FS_URI);
+    Configuration conf = new Configuration();
+    // Add all job configuration properties so they are picked up by Hadoop
+    for (String key : properties.getPropertyNames()) {
+      conf.set(key, properties.getProp(key));
+    }
+    this.fs = FileSystem.get(URI.create(uri), conf);
+
+    // initialize staging/output dir
+    this.stagingFile = new Path(WriterUtils.getWriterStagingDir(properties, numBranches, branchId), fileName);
+    this.outputFile = new Path(WriterUtils.getWriterOutputDir(properties, numBranches, branchId), fileName);
+    properties.setProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FINAL_OUTPUT_PATH, branchId),
```

Comment:
I tried using `WRITER_FILE_PATH` instead of `WRITER_FINAL_OUTPUT_PATH` but some of your tests were failing, so I'm guessing that was not the intended use for `WRITER_FILE_PATH`. Thats why I created a brand new key to track this.